### DISCUSSION TITLE
Installation/Requirements SMP

### DISCUSSION THREAD

**Discussion by sjaiser** [phax/phoss-smp#166] [Q&A]
Hello all,
I'm in the process of finding out what I need for a new installation of an SMP and found some informations, but am missing some others.

Are there recommendations on what to use as Backend Storage? XML/MySQL/MondoDB
And especially unter what circumstances might one be better than the other?

Are there recommendations on the setup regarding the needed uptimes for an SMP within the Peppol Network?
MySQL HA or maybe a K8S Setup to run the docker container to get a higher uptime?

Are there recommendations regarding the sizing of the systems where the Backend Storage (XML/MySQL/MongoDB) and the docker container is running?

Is there a document that I missed thay describes an installation?

What are the methods to update the SMP with new informations - can this only be done on the GUI or also on a cli?

Thanks a lot!
Steffen Jaiser

**Comment by phax** [comment]
Hi Steffen,
Thanks for asking these questions.
The different backend systems depend on your flavour of data storage - there are no big drawbacks for each of them.
* The XML backend is the most consistent one - writing all data as files on disk
* The SQL backend exists, because some IT departments are afraid of files - it writes all SMP specific data via SQL, some parts are still on the file system (yes, I know it is bad - historic reasons)
* The MongoDB backend exists for all the cool new hipster companies that don't like SQL anymore ;-) - it writes all SMP specific data (and a bit more) to MongoDB - still some elements remain on the file system (yes, I still know it is bad - historic reasons)

The requirements for Peppol are 99,5% - usually that is easily achieved with a single instance - but it of course depends on the number of participants you want to manage. I like the XML Backend because it is lightweight, consistent, easy to backup and easy to migrate - the only problem is - you cannot horicontally scale it, because the file systen data is "per-instance" only.
So there is no clear recommendation, as it heavily depends on your usage scenario. The system has been tested with 1mio entries in MongoDB and it worked like a charme.

Further documentation on how to configure and run an SMP can be found in the Wiki at https://github.com/phax/phoss-smp/wiki

It has a GUI as well as a REST API to fill it with content. See also in the Wiki for details.

hth, Philip

**Reply by sjaiser** [reply]
Hi Philip,
thanks a lot for the answer!
Just a few more questions...
Can you give a rough estimate on how big a xml Instance could grow with a reasonably sized system? Maybe 100K or 500K entries?
Do you need a short downtime if entries are updated or added or can this be done online (in all three different backends)? 
Could the XML Backend be run on a distributed filesystem, with more than one system accessing the data (maybe even in readonly mode) so in other words do the normal "Lookup" processes (besides adding/deleting/changing entries) write into the XML data?
Thanks a lot
Steffen

**Reply by phax** [reply]
Hi Steffen,
* all operations are performed on the fly - downtimes are only needed for version updates.
* Nope, running the XML backend  with a distributed file system does not work, because internally all data is loaded in memory on application startup and only changes are written to disk. That means if you run 2 instances, they would overwrite the changes of each other.
* The read process itself does not alter any stored data. It only creates log entries, but that is part of the application server.
* The size of an instance is hard to estimate, because it depends on the number of documents, on the size of the Business card etc. But usually we are talking about a few kilobytes per participant. For my http://smp.helger.com/ instance per today (5 participants, 1 Business Card) here a screendump with the file sizes: 
![grafik](https://user-images.githubusercontent.com/196219/123108726-f2b3e680-d43a-11eb-8185-b7b1df19c6ae.png)

hth, Philip

**Reply by sjaiser** [reply]
Hi Philip,
thanks a lot!
Best reagrds
Steffen

**Comment by Dataporti** [comment]
Hello Philip,

we are using Phoss SMP 8.0.12 with MariaDB 10.6. 
We plan to upgrade to MariaDB 11.4 in the future.

Do you foresee any potential compatibility issues?
We expect to it works. 

Thank you in advance for your reliable support.


Kind regards, 
Dataporti

**Reply by phax** [reply]
Hi @Dataporti, as MariaDB is not part of the officially supported RDMS systems, I can only say "I assume" so. My recommendation is, that you do upfront tests if everything works as expected. As usual, there is no guarantee ;-)
